packages/style-engine/class-wp-style-engine.php

```diff
@@ -293,6 +278,154 @@ protected static function is_valid_style_value( $style_value ) {
 		return true;
 	}
 
+	/**
+	 * Private constructor to prevent instantiation.
+	 */
+	private function __construct() {
+		// Register the hook callback to render stored styles to the page.
+		static::render_styles( array( __CLASS__, 'process_and_enqueue_stored_styles' ) );
```

Comment:
In terms of when we call this, at the moment the registration appears to happen when `get_instance()` happens to first be called in the code.

I was wondering if it's worth exploring using an explicit action for registration, so that it's predictable when the registration happens? I can't currently imagine a particular circumstance where this would be an issue, but something on my mind is the bug we fixed in https://github.com/WordPress/gutenberg/pull/42756 which was a difficult to diagnose one due to not really knowing when something was called.

I don't think this is a blocker for this PR, as it'd be easy enough for us to try out a different place for registration in a follow-up, but just thought I'd jot down the thought!

Reply:
> I was wondering if it's worth exploring using an explicit action for registration, so that it's predictable when the registration happens? 

It's a great question.

I was actually wondering whether the style engine should be enqueuing scripts/registering hooks at all, or whether Gutenberg (the consumer) should take care of its own registration / callbacks, and decides which stores it prints. 

I agree that it's not very clear how this PR registers the hook. The constraint was I wanted to only register it once, so  it was the easiest place to put it.

Reply:
> The constraint was I wanted to only register it once, so it was the easiest place to put it.

Exactly, I think that makes for a good argument for keeping it there for now, and we can look at where an ideal place is potentially in a follow-up?

Reply:
> something on my mind is the bug we fixed in https://github.com/WordPress/gutenberg/pull/42756 which was a difficult to diagnose one due to not really knowing when something was called.

Yeah, I've made it fairly opaque. It's not clear that `WP_Style_Engine::get_instance();` needs to be called for the styles to be enqueued.

That needs to be fixed I reckon, otherwise it might create a huge rabbit hole.